src/sage/rings/species.py

```diff
@@ -0,0 +1,1027 @@
+            sage: G.disjoint_direct_product_decomposition()
+            {{1, 2}, {3, 4}}
+            sage: G in C
+            False
+        """
+        if parent(G) == self:
+            return True
+        return len(G.disjoint_direct_product_decomposition()) <= 1
+
+    def _repr_(self):
+        r"""
+        Return the string representation for ``self``.
+
+        TESTS::
+
+            sage: from sage.rings.species import ConjugacyClassesOfDirectlyIndecomposableSubgroups
+            sage: C = ConjugacyClassesOfDirectlyIndecomposableSubgroups(); C
+            Infinite set of conjugacy classes of directly indecomposable subgroups
+        """
+        return "Infinite set of conjugacy classes of directly indecomposable subgroups"
+
+    Element = ConjugacyClassOfDirectlyIndecomposableSubgroups
+
+
+class AtomicSpeciesElement(Element):
+    def __init__(self, parent, dis, domain_partition):
+        r"""
+        Initialize an atomic species.
+
+        INPUT:
+
+            - ``dis``, a :class:`ConjugacyClassOfDirectlyIndecomposableSubgroups`
+
+            - ``domain_partition``, a dict representing the
+              assignment of each element of the domain of ``dis`` to
+              a "variable".
+
+        TESTS::
+
+            sage: A = AtomicSpecies(1)
+            sage: G = PermutationGroup([[(1,3),(4,7)], [(2,5),(6,8)], [(1,4),(2,5),(3,7)]])
+            sage: TestSuite(A(G)).run()
+        """
+        # Notice that a molecular species (and therefore an atomic species)
+        # must be centered on a multicardinality, otherwise it wouldn't be
+        # molecular. So it is kind of redundant to say "atomic species
+        # centered on a given multicardinality."
+        Element.__init__(self, parent)
+        self._dis = dis
+        self._dompart = domain_partition
+        self._mc = tuple(len(v) for v in self._dompart)
+        self._tc = sum(self._mc)
+
+    def _element_key(self):
+        r"""
+        Return the cache lookup key for ``self``.
+        """
+        return self._mc, self._dis
+
+    @cached_method
+    def _canonicalize(self):
+        r"""
+        Canonicalize this atomic species by sorting the orbits by
+        length and making them consecutive.
+
+        EXAMPLES::
+
+            sage: At = AtomicSpecies(2)
+            sage: G = PermutationGroup([[(1,2),(3,4),(5,6),(7,8,9,10)]]); G
+            Permutation Group with generators [(1,2)(3,4)(5,6)(7,8,9,10)]
+            sage: H = PermutationGroup([[(1,2,3,4),(5,6),(7,8),(9,10)]]); H
+            Permutation Group with generators [(1,2,3,4)(5,6)(7,8)(9,10)]
+            sage: A = At(G, {1: [1,2,3,4], 2: [5,6,7,8,9,10]})
+            sage: A._dompart
+            ((5, 6, 7, 8), (9, 10, 1, 2, 3, 4))
+            sage: C = At(G, {1: [1,2,5,6], 2: [3,4,7,8,9,10]})
+            sage: C._dompart
+            ((5, 6, 7, 8), (9, 10, 1, 2, 3, 4))
+        """
+        # The canonicalization is done in the element constructor.
+        pass
```

Comment:
If we canonicalize the group, then the domain partition also needs to be updated. I can't do this after getting the cached group because it is already canonical and I can't get the required transformation on the domain partition.